rspirv/sr/autogen_context.rs

```diff
@@ -0,0 +1,408 @@
+    pub fn type_runtime_array(&mut self, element_type: Token<Type>) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::RuntimeArray { element_type },
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_opaque(&mut self, type_name: String) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::Opaque { type_name },
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_pointer(
+        &mut self,
+        storage_class: spirv::StorageClass,
+        pointee_type: Token<Type>,
+    ) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::Pointer {
+                storage_class,
+                pointee_type,
+            },
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_function(
+        &mut self,
+        return_type: Token<Type>,
+        parameter_types: Vec<Token<Type>>,
+    ) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::Function {
+                return_type,
+                parameter_types,
+            },
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_event(&mut self) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::Event,
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_device_event(&mut self) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::DeviceEvent,
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_reserve_id(&mut self) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::ReserveId,
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_queue(&mut self) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::Queue,
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_pipe(&mut self, qualifier: spirv::AccessQualifier) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::Pipe { qualifier },
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_forward_pointer(
+        &mut self,
+        pointer_type: Token<Type>,
+        storage_class: spirv::StorageClass,
+    ) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::ForwardPointer {
+                pointer_type,
+                storage_class,
+            },
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_pipe_storage(&mut self) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::PipeStorage,
+            decorations: Vec::new(),
+        })
+    }
+    pub fn type_named_barrier(&mut self) -> Token<Type> {
+        self.types.fetch_or_append(Type {
+            ty: TypeEnum::NamedBarrier,
+            decorations: Vec::new(),
+        })
+    }
+    pub fn lift_extension(
```

Comment:
I'm not sure that these `lift_*` functions belong to the `Context`. I think these `lift_*` functions are only useful for converting a DR module into a SR module. But `Context` is long-lived and intended to be the central hub for everything that need to be uniqued. Should we separate the concerns and have a `Lifter` class or similar for the conversion logic?

Reply:
I think I just not fully understand the concept behind `Context`. Is the idea for the user to keep it between multiple modules being loaded from DR? How would this be beneficial as opposed to just starting from a new `Context` for every module being loaded?

Reply:
the lifting instructions would need access to `Storage` structs that only `Context` has, that's why I was putting them into `Context`.

Reply:
No, one `Context` is just for one module. A `Contenxt` is the owner of all module's contents: types, constants, instructions, etc. It maintains the lifetime of all those components. `Context` is MR specific here. My previous commit is about layering. Those `lift_*` methods transforms DR into MR and that's a one-time effort. So it's a bridge between DR and MR. Theoretically MR should not know about DR because they are at different layers. That mans `Context`, which belongs to MR, should not have functions only for DR. It should only contain functions for maintaining MR constructs. Another way to say this, if we have another input format that can be converted into MR, do we also put the conversion functions into the `Context` and requiring users of MR to also depend on the input format even if they don't use the input format at all? Does this make sense? 

I guess it's fine if you have internal dependencies so `lift_*` functions must be put in `Context` right now. I don't want to let you rebase the whole chain by making big changes on this one. But I think we should think about how to properly layer this. Could you leave a TODO here and get back to it later? Thanks!

Reply:
Thank you for explanation! I'm a great deal confused by "MR" in your comment. Perhaps, you meant "SR"? I hope it's not an intro to another grand rename of things :)

I feel like SR may not need any kind of context at all. I mean, possibly it will. But today it seems simpler to just store `Storage` things directly in the `Module` which is being populated, since they will likely stay there and be exposed to the user at some point. I'd like this PR to *not* be one that decides the fate of `Context`.

That is to say, I agree that separating the DR -> SR logic ("lifting") from other things in SR makes up for a good abstraction. We could, for now, just restrict the functions to `pub(in crate::sr)` and figure out where to put them later. We could even declare a trait with these functions that `Context` implements, but that would be going too far.